### PR TITLE
player: fix problems in sort_armor

### LINKED ISSUE #42503: Removing a “survivor headlamp (on)” from a follower through the “r Sort armor” menu causes a crash/segmentation fault. EDIT: there's the 0F temperature bug as well.
# Describe the bug
Removing a “survivor headlamp (on)” from a follower through the “r Sort armor” menu causes a crash/segmentation fault.

# Steps To Reproduce
1. Wake or don’t wake the follower.
2. Move into the follower.
3. Choose “r Sort armor”
4. Select the “survivor headlamp (on)”
5. Press u
6. Press y

# Expected behavior
The game should not crash when removing a followers’ worn items.

# Versions and configuration
- OS: Linux
    - OS Version: LSB Version: n/a; Distributor ID: ManjaroLinux; Description: Manjaro Linux; Release: 20.0.3; Codename: Lysia; 
- Game Version: 0.E-4458-g8cf46cdb44 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    No Fungal Monsters [No_Fungi],
    Fuji's Military Profession Pack [fuji_mpp],
    SpeedyDex [speedydex],
    Stats Through Kills [stats_through_kills],
    Stats Through Skills [StatsThroughSkills]
]

# Additional context
Debug and crash logs [cataclysm-dda.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4988706/cataclysm-dda.zip)
Save file [Alta Vista.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4988712/Alta.Vista.zip) this is "a save file that occurs in the right conditions, just before the crash happens."
To take off the “survivor headlamp (on)” from the follower without causing a crash you need to remove everything else before removing the survivor headlamp.
Alas trying to remove the military rucksuck that contains the "survivor headlamp (on)” also causes a crash.

### LINKED ISSUE #42508: Sort Armor menu not showing coverage or encumbrance in the middle
# Describe the bug

In the middle part of the Sort Armor menu, coverage and encumbrance always show up as 0 no matter the piece of clothing selected. The encumbrance on the right part is however correct. The items' descriptions also correctly note the coverage and encumbrance.

# Steps To Reproduce

1. Have clothing, for instance a cassock.
2. Open "Re-layer Armor/Clothing" menu.
3. Scroll through all your clothing and observe how coverage and encumbrance stay at 0.

# Expected behavior

The menu should be showing everything properly.

# Screenshots

image
With the Aftershock mod installed as per Versions and configuration below.

image
Different world, same version, mods are also the same except no Aftershock. As you can see the bug is not caused by Aftershock's presence.

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 1903
- Game Version: 0.E-4463-g344d6de [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Graphical Overmap [Graphical_Overmap],
    Aftershock [aftershock]
]

### PR DESCRIPTION
**Pull request by andrei8l**
#### Summary
`SUMMARY: Bugfixes "Update and fix problems in player::sort_armor"`

#### Purpose of change
Unequipping last item from the sort armor menu crashes the game
Middle pane doesn't show encumbrance/coverage for "All" category
`sort_armor()` uses hardcoded body part list

fixes #42503
fixes #42508

#### Describe the solution
Don't crash game
Show average encumbrance & coverage for the "All" category
Use the character's actual body parts

#### Describe alternatives you've considered
N/A

#### Testing
Sort armor menu doesn't crash game for any interaction and displayed values are sensible

#### Additional context
The replacement for `combine_bp` (to combine body parts if they are covered by the same clothing) is pretty ugly... sorry
@Fris0uman review pls

### CODE REVIEW

File: src/armor_layers.cpp [hunk 1]
```diff
@@ -86,7 +86,7 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
     std::vector<std::set<std::string>> lists_of_bad_items_within;
 
     for( const bodypart_id &bp : c.get_all_body_parts() ) {
-        if( bp->token != tabindex && num_bp != tabindex ) {
+        if( bp != _bp && _bp != bodypart_id( body_part::num_bp ) ) {
```
Comment (Fris0uman) [inline src/armor_layers.cpp:89]:
```suggestion
        if( bp != _bp && _bp != bodypart_id( "num_bp" ) ) {
```

File: src/armor_layers.cpp [hunk 2]
```diff
@@ -295,11 +295,16 @@ std::vector<std::string> clothing_properties(
     props.reserve( 5 );
 
     const std::string space = "  ";
+
+    const int coverage = bp == bodypart_id( body_part::num_bp ) ? worn_item.get_avg_coverage() :
```
Comment (Fris0uman) [inline src/armor_layers.cpp:299]:
```suggestion
    const int coverage = bp == bodypart_id( "num_bp" ) ? worn_item.get_avg_coverage() :
```

File: src/armor_layers.cpp [hunk 3]
```diff
@@ -295,11 +295,16 @@ std::vector<std::string> clothing_properties(
...
+                         worn_item.get_coverage( bp );
+    const int encumbrance = bp == bodypart_id( body_part::num_bp ) ? worn_item.get_avg_encumber(
```
Comment (Fris0uman) [inline src/armor_layers.cpp:301]:
```suggestion
    const int encumbrance = bp == bodypart_id( "num_bp" ) ? worn_item.get_avg_encumber(
```

File: src/armor_layers.cpp [hunk 4]
```diff
@@ -439,8 +444,17 @@ void player::sort_armor()
     * because some items can have multiple entries (i.e. cover a few parts of body).
     */
 
-    int req_right_h = 3 + 1 + 2 + num_bp + 1;
-    for( const bodypart_id &cover : get_all_body_parts() ) {
+    // FIXME: get_all_body_parts() doesn't return a sorted list
+    //        and bodypart_id is not compatible with std::sort()
+    //        so let's use a dirty hack
+    cata::flat_set<bodypart_id> armor_cat;
+    for( const auto &it : get_all_body_parts() ) {
```
Comment (Fris0uman) [inline src/armor_layers.cpp:451]:
```suggestion
    for( const bodypart_id &it : get_all_body_parts() ) {
```

File: src/armor_layers.cpp [hunk 5]
```diff
@@ -439,8 +444,17 @@ void player::sort_armor()
...
+        armor_cat.insert( it );
+    }
+    armor_cat.insert( bodypart_id( body_part::num_bp ) );
```
Comment (Fris0uman) [inline src/armor_layers.cpp:454]:
```suggestion
    armor_cat.insert( bodypart_id( "num_bp" ) );
```
isn't num_bp already returned by get_all_body_parts()?

File: src/armor_layers.cpp [hunk 6]
```diff
@@ -547,14 +555,14 @@ void player::sort_armor()
     ui.on_redraw( [&]( const ui_adaptor & ) {
         // Create ptr list of items to display
         tmp_worn.clear();
-        if( tabindex == num_bp ) {
+        const bodypart_id bp = armor_cat[ tabindex ];
```
Comment (Fris0uman) [inline src/armor_layers.cpp:558]:
```suggestion
        const bodypart_id &bp = armor_cat[ tabindex ];
```

File: src/armor_layers.cpp [hunk 7]
```diff
@@ -646,32 +655,34 @@ void player::sort_armor()
...
         } else if( rightListOffset + rightListLines > rightListSize ) {
             rightListOffset = rightListSize - rightListLines;
         }
-        for( int cover = 0, pos = 1, curr = 0; cover < num_bp; cover++ ) {
-            bool combined = false;
-            if( combine_bp( cover ) ) {
-                combined = true;
+        int pos = 1, curr = 0;
+        for( const bodypart_id cover : rl ) {
+            if( cover == bodypart_id( body_part::num_bp ) ) {
```
Comment (Fris0uman) [inline src/armor_layers.cpp:680]:
```suggestion
            if( cover == bodypart_id( "num_bp" ) ) {
```

**Comment by andrei8l** [comment]
Thanks a lot for the quick review!

**Pull request merged**